comp/rdnsquerier/impl/rdnsquerier.go

```diff
@@ -181,6 +183,63 @@ func (q *rdnsQuerierImpl) GetHostname(ipAddr []byte, updateHostnameSync func(str
 	return nil
 }
 
+// GetHostnameSync attempts to resolve the hostname for the given IP address synchronously.
+// If the IP address is invalid then an error is returned.
+// If the IP address is not in the private address space then it is ignored - no lookup is performed and nil error is returned.
+// If the IP address is in the private address space then the IP address will be resolved to a hostname.
+// The function accepts a timeout duration, which defaults to 2 seconds if not provided.
+func (q *rdnsQuerierImpl) GetHostnameSync(ctx context.Context, ipAddr string) (string, error) {
```

Comment:
I have the feeling that things could be simpler for a  **synchronous** reverse DNS (with cache).

Maybe the need for chan/mutex/context timeout is due to initial implementation being **very async oriented**.

Reply:
What I mean is that, it seems that for a **synchronous** reverse DNS with cache, then basic logic could be something like:
- client request dns entry
- the reverse DNS library checks if entry is in cache and returns results
- if not, make actual reverse DNS calls, add results to cache and returns results
with everything above being synchronous

But instead, we have much complex setup:
- client request dns entry
- the reverse DNS library checks if entry is in cache and callback is called
- if not:
  - push request to a queue (`querierImpl.getHostnameAsync`)
  - queue is consumed by workers
  - workers make actual reverse DNS calls
  - results are added to cache and callbacks are called
- callback/chan/mutex are needed in GetHostnameSync to interact with the async GetHostname

Reply:
Hum, maybe having a underlying implementation can be beneficial for this: https://github.com/DataDog/datadog-agent/pull/30002/files#r1814405433

Reply:
Yeah, I've thought about that as well. I think there are some benefits to leveraging at least some of the whole workers paradigm.

Consider a large server with Network Traffic Paths enabled. A large number of workers is required to handle this load. With the first implementation described above, we're likely to make many more calls to the DNS server as the parallelism scales up. Depending on the traffic patterns we detect and the user's network topology, this might be more of a cold start problem than a continuous issue.

When @jmw51798, @vicweiss, and I discussed it getting all of that stuff to work with a synchronous implementation seemed like a large effort.